Save user coordinates on local storage and reuse them if possible

ranyitz opened the pull request:
Fixes #30

ranyitz committed:
save user coordinates on local storage and reuse them if possible

ranyitz committed:
small fixes

uriklar commented on src/App.js:
```diff
@@ -158,7 +167,11 @@ function App() {
                 coordinates={state.userCoordinates}
-                setCoordinates={(coords) => dispatch({ type: 'setUserCoordinates', payload: coords })}
+                setCoordinates={(coords) => {
+                  // Save the user coordinates in order to reuse them on the next user session
+                  setUserCoordinatesToLocalStorage(coords);
```
I'd put this in the handler for the setUserCoordinates event in the reducer. That way it will be called even if the event is dispatched from someplace else

ranyitz replied:
I thought about that as well, and putting it in the reducer creates a side-effect in the reducer which is generally something I prefer to avoid. You can create an action that both does the side-effect and fire the event. Specifically in this case, it creates a situation where we set the coordinates to local storage just after we retrieved them from local storage, which is a noop but still doesn't make a lot of sense.

uriklar commented on src/localStorage.js:
```diff
@@ -0,0 +1,7 @@
+export const getUserCoordinatesFromLocalStorage = () => {
+    return JSON.parse(localStorage.getItem('user_coordinates'));
```
Might be a nitpick but i'd use a more specific name "ONEKM__user_coordinates" or something..

ranyitz replied:
I don't mind changing it but, local storage is already unique per domain, I don't see a reason to use namespacing here.

uriklar commented on src/App.js:
```diff
@@ -51,6 +52,14 @@ function App() {
   const [state, dispatch] = useReducer(reducer, initialState);
 
   useEffect(() => {
+    if (!state.userCoordinates) {
+      const cachedCoordinates = getUserCoordinatesFromLocalStorage();
+
+      if (cachedCoordinates) {
+        dispatch({ type: 'setUserCoordinates', payload: cachedCoordinates })
+      }
+    }
```
This useEffect is getting really busy, which makes it hard to maintain and bug prone. 
Can we separate this code to a separate use effect that will be called on mount?

uriklar reviewed (changes requested):
Hi @ranyitz , thank you very much!
I requested some small changes.

uriklar committed:
separate use effect
don't show modal if has coordinates
add eslint ignore to disable autofix for exhaustive hooks error

uriklar committed:
move localStorage set to reducer handler

uriklar commented:
@ranyitz thanks I added the small changes + made the modal not appear when user has saved coordinates

ranyitz commented:
@uriklar Thanks for reviewing the PR. To my understanding, the PR is ready for merge. let me know if there is anything left to do on my end.

guytepper committed:
Merged master and changed method names

guytepper committed:
Merge branch 'master' into cache-user-coordinates

guytepper commented:
Hi @ranyitz ! Sorry for the delay.
Thank you for the suggestion & implementation. It works great!
Note that I changed the `localStorage` methods to be more generic.

Also, we got this report an hour ago:

<img width="380" alt="Screen Shot 2020-10-08 at 1 19 59" src="https://user-images.githubusercontent.com/13344923/95394075-63928d00-0904-11eb-951c-126c5acc4bb8.png">

You just made someone's life better today 🙏
